Refactor tests

amroel opened the pull request:
A gross description of the changes in this pr...

Eliminated and replaced IntegrationTestOptions with 2 file folders: 
AvailableTestConfigs holding simple xml files with all what was in IntegrationTestOptions and CurrentTestConfig holding a "TestConfig.xml" which is intended to configure a database the tests should run against.
If a test is intended to run against a specific database that is not configured it will be ignored (not fail!) otherwise it will run.
Manual steps To configure a database: 
1. find the corresponding xml configuration file in "AvailableTestConfigs"
2. copy that file to "CurrentTestConfig\TestConfig.xml". Create "CurrentTestConfig" folder if necessary.
3. adjust the connectionString in TestConfig.xml to fit your needs.

Those changes should allow different build configurations on a build server to run the integration tests for each possible db. As far i can tell that would be: SQLite, SqlCe and Firebird embedded.

I've run all the tests on my local machine also against SqlServer and it looks good. 
I didn't test against all other Databases!

Known "Issues":
When running against Firebird embedded or SqlCe the native db libraries must be copied manually to the test folder (src\FluentMigrator.Tests\bin\Debug\). On a build server this can be easily done by a script.

escarls commented:
Build fails on Travis and two tests fails on TeamCity.

amroel commented:
@escarls regarding TC Build: I'll take a look as soon as possible. It must be something trivial that I overlooked.

Travis is a complete different story...
`Target PostBuildEvent:
Executing: copy /home/travis/build/fluentmigrator/fluentmigrator/lib/SQLite.Interop.dll /home/travis/build/fluentmigrator/fluentmigrator/src/FluentMigrator.Tests/bin/Debug/
	/tmp/tmp76de4268.tmp: 1: /tmp/tmp76de4268.tmp: copy: not found
/usr/lib/mono/4.0/Microsoft.Common.targets: error : Command 'copy /home/travis/build/fluentmigrator/fluentmigrator/lib/SQLite.Interop.dll /home/travis/build/fluentmigrator/fluentmigrator/src/FluentMigrator.Tests/bin/Debug/' exited with code: 127.
Task "Exec" execution -- FAILED
`

First, part of the changes is a PostBuildEvent in FluentMigrator.Tests.csproj that looks like this: [https://github.com/fluentmigrator/fluentmigrator/pull/784/commits/6cbafbc9d6f056a09fbef2977e0288089076e873](url)
The step is obviosly wrong in a linux environment! I'm not sure what to do here and so I need help.

Second, I dont't think that running the **integaration** tests in a linux environment would be working at all:
1. The above mentioned SQLite.Interop is a native windows library
2. The native SqlCe libraries are windows libraries
3. The currently used firebird embedded libraries are windows libraries

Third, and this is just a side note: Why do have two build servers?

fubar-coder commented:
Teamcity is Windows, while Travis is Linux

amroel commented:
@fubar-coder I am aware that travis is linux and teamcity is windows but still...
I mean travis is only a build server just as good or bad as any other build server.
*Building* the project in a linux environment doesn't mean that it is also *runnable* in that environment.
But again...that was just a side note :)

escarls commented:
@amroel the firebird libs are included in Linux that part worked before your change ;)

amroel commented:
I'm going to revert PostBuildEvent step.
Instead I'm going to write a batch file or powershell script that would create the folder "CurrentTestConfig", copy the desired configuration file to it and in case of firebird, SQLite and SqlCe copy the native libraries to the test folder.
Developers would run the scripts locally on their machines.

The script should be runnable on Teamcity so that no further actions are required there.
On Travis we would need to define a separate build for running the tests against firebird.

Thoughts?

escarls commented:
@amroel I found some good advice for the build events [here](http://www.mono-project.com/archived/porting_msbuild_projects_to_xbuild/), when I changed from PostBuildEvent to Target AfterBuild, it builds, but fails when running sql lite tests. I assume it's like you said, that I'd need the native SQLLite dll for Linux to run that. I'll do some more testing, [here](https://travis-ci.org/escarls/fluentmigrator/builds/209488334) you can see the log from my test build with just the switch from PostBuildEvent to Target

amroel commented:
@escarls aaah that's nice.
Should I also do that and push? What do you think about my last suggestion?

fubar-coder commented:
This is a huge set of changes and it will take some time to review it. Can you in detail describe the following questions (some for documentation purpose only):

1. What were the structural changes to the test infrastructure?
2. What changes where done in which project, and why?
3. What are the consequences for the build infrastructure?
4. Did you add new package references or did you update some?

Answering those questions would help me understand your PR and why you did some things the way you did.

amroel commented:
1, What were the structural changes to the test infrastructure?
- Added a TestFixtureSetup to IntegrationTestBase. This triggers the configuration of a database.
- Added a TestFixtureTearDown to IntegrationTestBase. This should cleanup after a suits of tests has been run (like dropping a database).
- Renamed ExecuteWithSupportedProcessors to ExecuteWithSupportedProcessor in IntegrationTestBase to reflect the fact that we're not running a test for all supported processors at the same time.
- Dropped the class IntegrationTestOptions.
- Created a CreateProcessor method in IntegrationTestBase. This is used by subclasses.
- Every test class in the migration namespace ultimately inherits inherits IntegrationTestBase using its methods and properties to figure out if a test should run and, if yes, use the appropriate processor.

2. What changes where done in which project, and why?
Nearly all the changes are in the test project. But after successfully running all tests I found out that some tests were **meaningfully** failing for firebird and SqlCe so I fixed that.
(I expect that to be the case for all other databases except SqlServer! Those tests were nearly never executed!)

3. What are the consequences for the build infrastructure?
I'm not really sure if I understand your question. The whole point of this PR is to be able to run the integration tests on the build server(s)!

4. Did you add new package references or did you update some?
The libraries for SqlLite was so outdated that the tests wouldn't run due an exception that the referenced assembly cannot run under .net 3.5!
That's the only package I had to update. The updated System.Data.SQLite.dll requires SQLite.Interop.dll which was not the case before.

I hope this might help you and others more.

amroel commented:
sorry, i accidentally closed the PR and had to reopen it again :)

fubar-coder commented:
Thank you for the explanation. This will help during the review.
